### ISSUE TITLE
star.exe goes into an infinite loop if one of the restarted apps throws an exception

### ISSUE THREAD

**Issue by warpech** [Starcounter/Home#458]
star.exe goes into an infinite loop if one of the restarted apps throws an exception.

#### 1. Create file test.cs

```cs
class Program
{
    static void Main(string[] args)
    {
        
    }
}
```

#### 2. Create file test2.cs

```cs
using System;
using System.IO;

class Program
{
    static void Main(string[] args)
    {
        if(File.Exists("c:\\empty.txt")) 
        {
            throw new Exception("I refuse to start when c:\\empty.txt exists");
        }
    }
}
```

#### 3. Run two apps:

```
star test.cs
star test2.cs
```

#### 5. Restart the first app

```
star test.cs
```

## Expected

The app `test.cs` should restart. The app `test2.cs` should throw an exception and break the startup  procedure.

## Actual

The apps go into an infinite loop of restarting:

```
test -> default
test -> default (restarting default (and 1 other executable(s)))
test -> default (restarting default (and 1 other executable(s)) ..)
test -> default (restarting default (and 1 other executable(s)) ...)
test -> default (restarting default (and 1 other executable(s)) .)
test -> default (restarting default (and 1 other executable(s)) ..)
test -> default (restarting default (and 1 other executable(s)) ...)
(...)
```

**Comment by miyconst** [comment]
@alemoi could you take a look please?

@bigwad could be it related to your latest app startup changes?

**Comment by warpech** [comment]
Moreover, after it goes into the infinite loop, `staradmin stop db` can't stop the database. Only `staradmin kill all` helps.

**Comment by alemoi** [comment]
I have started the investigation and it seems there are some hanging mechanisms in `scadminserver`:

![image](https://user-images.githubusercontent.com/208134/51485282-f9d3f780-1d9d-11e9-8023-0c0dfcc9316b.png)

I am not aware of how exactly the app management is supposed to work in `scadminserver`. I guess the best one to check this is @per-samuelsson 

My guess would be that the `scadminserver` is waiting for some signal from restarting sccode, which is already dead at that point and not sending anything back to `scadminserver`.

> The apps go into an infinite loop of restarting:

How do you know that apps are restarting? Because I don't see any `sccode` process after the `star test.cs` is run second time. I see a hanging `scadminserver` instead.
Its also interesting that restarting second time the exception app `star test2.cs` - it does not cause this hang.

**Comment by per-samuelsson** [comment]
I can help diagnosing this if no one else does that.

Do we know if this is a new flaw, and this use case has worked before?

**Comment by per-samuelsson** [comment]
FYI, I've set it up locally and using latest sources, and I can repro it.

**Comment by per-samuelsson** [comment]
OK, so I found it's [`HttpInternal.POST` here](https://github.com/Starcounter/level1/blob/develop/src/Server/Starcounter.Server/Commands/Processors/StopExecutableCommandProcessor.cs#L148-L149) that is used to call from the admin server into the code host, telling it to load an app and return it's entrypoint, that never return in case the host process dies (which is what happen).

Have a vague memory we've had some similar problem in the past: a call is made using our REST API, and routed to a handler in a foreign process that terminates, and that call then never return.

I guess I can redesign the admin server to work around it, but want to hear what @alemoi has to say about it first. Is this behavior the expected? Do you also remember we've had something like this in the past?

**Comment by alemoi** [comment]
> OK, so I found it's HttpInternal.POST here that is used to call from the admin server into the code host, telling it to load an app and return it's entrypoint, that never return in case the host process dies (which is what happen).

I guess we just need to have a timeout on this POST REST call (`Http` API allows setting max response time), because host can occasionally die by some other means.

**Comment by alemoi** [comment]
> So you don't think it's worth fixing it in the Http API itself, so we don't hit more of this in other spots?

No, thought only about this particular call.
But its super-low prio I would say, isn't it?

**Comment by warpech** [comment]
> How do you know that apps are restarting? Because I don't see any sccode process after the star test.cs is run second time. I see a hanging scadminserver instead.
Its also interesting that restarting second time the exception app star test2.cs - it does not cause this hang.

All I know is that `star.exe` writes a new line with the word "restarting" every second, infinitely:

```text
test -> default
test -> default (restarting default (and 1 other executable(s)))
test -> default (restarting default (and 1 other executable(s)) ..)
test -> default (restarting default (and 1 other executable(s)) ...)
test -> default (restarting default (and 1 other executable(s)) .)
test -> default (restarting default (and 1 other executable(s)) ..)
test -> default (restarting default (and 1 other executable(s)) ...)
(...)
```

**Comment by miyconst** [comment]
Closing as won't fix in Starcounter 2.4 and not applicable in Starcounter 3.0.
